File: android/src/main/java/com/tal/d_stack/DStack.java
package com.tal.d_stack;

import android.content.Context;
import android.content.Intent;

import java.util.Map;

import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.embedding.engine.FlutterEngineCache;
import io.flutter.embedding.engine.dart.DartExecutor;

public class DStack {
    public static final String ENGINE_ID = "d_stack_engine";

    /*package*/ FlutterEngine flutterEngine;
    private boolean running;

    private DStack() {
    }

    private interface Holder {
        DStack instance = new DStack();
    }

    public static DStack getInstance() {
        return Holder.instance;
    }

    public void init(Context context) {
        if (!running) {
            DNavigationManager.getInstance().init(context);
            flutterEngine = new FlutterEngine(context);
            flutterEngine.getDartExecutor().executeDartEntrypoint(
                    DartExecutor.DartEntrypoint.createDefault()
            );
            FlutterEngineCache.getInstance().put(ENGINE_ID, flutterEngine);
            running = true;
        }
    }

    public interface NativeRoute {
        Intent build(Context context);
    }

    public void registerRoute(Map<String, NativeRoute> routeMap) {
        DNavigationManager.getInstance().registerRoute(routeMap);
    }

    public void pushRoute(String routeName) {
        DNavigationManager.getInstance().pushRoute(routeName);
    }

    public void pop() {
        DNavigationManager.getInstance().pop();
    }
}

File: ios/Classes/DStack.m
#import "DStack.h"
#import <UIKit/UIKit.h>
#import "DNavigationManager.h"

NSString * const EngineId = @"d_stack_engine";

@interface DStack()

@property (nonatomic) BOOL running;

@end

@implementation DStack

+ (instancetype)shared {
    static id instance = nil;
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

- (void)initialize {
    if (!self.running) {
        self.flutterEngine = [[FlutterEngine alloc] initWithName:EngineId];
        self.running = [self.flutterEngine run];
    }
}

- (void)registerRoute:(NSDictionary<NSString *,NativeRoute> *)routeMap {
    [DNavigationManager.shared registerRoute:routeMap];
}

- (void)pushRoute:(NSString *)routeName {
    [DNavigationManager.shared pushRoute:routeName];
}

@end
